CARDS-1561: On the patient identification screen, replace the date field with dropdowns for day,month,year

veronikaslc opened the pull request:
[CARDS-1561](https://phenotips.atlassian.net/browse/CARDS-1561): On the patient identification screen, replace the date field with dropdowns for day,month,year

Inspired by [react-dropdown-date](https://github.com/ssxv/react-dropdown-date) package.

marta- reviewed (dismissed):
@veronikaslc make two separate commits:
* one with the original file
* one with your changes to it

Commit:
CARDS-1561: On the patient identification screen, replace the date field with dropdowns for day,month,year
Original files from react-dropdown-date package

Commit:
CARDS-1561: On the patient identification screen, replace the date field with dropdowns for day,month,year

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+
+  let generateYearOptions = () => {
+    let yearOptions = [];
+    yearOptions.push(<MenuItem key={-1} value="-1">yyyy</MenuItem>);
```
"yyyy" looks awkward at the to of the list, I would remove the label completely.
```suggestion
    yearOptions.push(<MenuItem key={-1} value="-1"></MenuItem>);
```

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+
+  let generateMonthOptions = () => {
+    let monthOptions = [];
+    monthOptions.push( <MenuItem key={-1} value="-1">mm</MenuItem> );
```
```suggestion
    monthOptions.push( <MenuItem key={-1} value="-1"></MenuItem> );
```

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+
+  let generateDayOptions = () => {
+    let dayOptions = [];
+    dayOptions.push(<MenuItem key={-1} value="-1">dd</MenuItem>);
```
```suggestion
    dayOptions.push(<MenuItem key={-1} value="-1"></MenuItem>);
```

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+        autoFocus={autoFocus && order.indexOf("year") == 0}
+        renderValue={(selected) => {
+            if (selected < 0 ) {
+              return <div className={classes.default}>yyyy</div>;
```
Now that we're not tied to the native date picker, let's make the components more user friendly with human-readable placeholders.

```suggestion
              return <div className={classes.default}>Year</div>;
```

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+        autoFocus={autoFocus && order.indexOf("month") == 0}
+        renderValue={(selected) => {
+            if (selected < 0 ) {
+              return <div className={classes.default}>mm</div>;
```
```suggestion
              return <div className={classes.default}>Month</div>;
```

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+        autoFocus={autoFocus && order.indexOf("day") == 0}
+        renderValue={(selected) => {
+            if (selected < 0 ) {
+              return <div className={classes.default}>dd</div>;
```
```suggestion
              return <div className={classes.default}>Day</div>;
```

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+            if (selected < 9 ) {
+              return "0" + (selected+1);
+            }
+            return selected+1;
```
This should display the month name, not the number, for consistency with the dropdown items.

sashaandjic commented:
The calendar does not work. Should it be there at all?
![image](https://user-images.githubusercontent.com/23243244/149002874-3a45dcbf-c62a-4b7e-a7f7-a72179ca4de5.png)

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+  return (
+    <div id="dropdown-date" className={classes.container}>
+      { order.map(part => { return renderParts[part]() }) }
+      <CalendarTodayIcon fontSize="small" className={classes.icon}/>
```
This icon should be removed. it doesn't do anything and that makes its presence confusing.

marta- commented on modules/commons/src/main/frontend/src/components/DropdownsDatePicker.jsx:
```diff
@@ -0,0 +1,295 @@
+};
+
+const useStyles = makeStyles(theme => ({
+  select : {
+    paddingLeft: theme.spacing(1),
+  },
```
Let's remove the borders under each select, they look sloppy on top of the container's border. The hover effect (thick border) should be kept though.

Commit:
CARDS-1561: On the patient identification screen, replace the date field with dropdowns for day,month,year
* Replaced placeholders
* Removed icon and bottom border

sashaandjic commented:
One more small thing. Can we remove scroller from MRN field.

![image](https://user-images.githubusercontent.com/23243244/149024890-d740bc78-9ad2-4813-b165-ecc35402c460.png)

marta- committed:
CARDS-1561: On the patient identification screen, replace the date field with dropdowns for day,month,year
* Added support for full width dropdown date
* Improved "empty" options in each dropdown
* Refactored styling

marta- commented:
Rebased on dev. I also made the dropdown date picker stretch full width and reworked the empty optopns at the top of each dropdown in 29c51aa. @sashaandjic please retest.

marta- commented:
Would be nice to have: when selecting the year, the focus should automatically move to the month dropdown. Same for month -> day. A quick attempt to implement this was made in 4873f4b on the `CARDS-1561_focus` branch, but it's not working. @veronikaslc can you pick it up and see if you have more success with it? (low priority)

sashaandjic reviewed (approved):
looks good!
![image](https://user-images.githubusercontent.com/23243244/149571464-a7b41048-2d99-4ed8-b6a8-327b60039d0e.png)

sashaandjic commented:
I think making the details font smaller and centering it would look better.
![image](https://user-images.githubusercontent.com/23243244/149571801-1e5647c6-4b95-4531-801c-199c66fd3ee4.png)

veronikaslc commented:
@sashaandjic Do you mean "Date of birth" above the inputs on the left side?

sashaandjic commented:
this text 
![image](https://user-images.githubusercontent.com/23243244/149575119-c6fd2915-8fc2-4da1-92f7-3f8f47b71c7d.png)

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/149575440-497208f1-e6e7-4fe8-a963-d3c930505648.png)

marta- commented:
> I think making the details font smaller and centering it would look better

@sashaandjic Noted. I'll take care of it as part of https://phenotips.atlassian.net/browse/CARDS-1574. Please add to that task any other issues that you might find.

sashaandjic commented:
@marta-  in Terms of Use we say "DATA-PRO" and here we say "DATAPRO". Should I add this to 1574 as well?

marta- commented:
> @marta- in Terms of Use we say "DATA-PRO" and here we say "DATAPRO". Should I add this to 1574 as well?

No, that's https://phenotips.atlassian.net/browse/CARDS-1563.

sashaandjic commented:
> Would be nice to have: when selecting the year, the focus should automatically move to the month dropdown. Same for month -> day. A quick attempt to implement this was made in [4873f4b](https://github.com/data-team-uhn/cards/commit/4873f4bdfcb761f3c1911f0c198cf9ff4cb73285) on the `CARDS-1561_focus` branch, but it's not working. @veronikaslc can you pick it up and see if you have more success with it? (low priority)

Logged as improvement [CARDS-1575](https://phenotips.atlassian.net/browse/CARDS-1575)

acrowthe commented:
Nothing I'd consider critical, but I've got a couple UI/UX suggestions. All tested on Google Chrome for Ubuntu

Mouse controls it works great, but it doesn't work as nicely as I would hope using just keyboard:
- When you get focus on a field (year, month or day) using tab, typing in letter/numbers has no effect until you press enter to open the dropdown. I'd expect tabbing to the year field then pressing a number to open the dropdown and start selecting whatever is typed.
  - Expected functionality: Tab to month, type "Jan", dropdown opens up with January highlighted
  - Current functionality: Tab to month, press enter, type "Jan", January is highlighted
- When tabbing out of an open dropdown, the current selected option is not applied
  - Expected functionality: Have the month drop down open, with January highlighted. Press tab, January is filled into the month and focus switches to Day
  - Current functionality: Have the drop down open, with January highlighted. Press enter to select January, then tab to switch to Month
- When trying to type in a year, it seems that it only considers the first two entered numbers
  - Expected functionality: With the year drop down open, typing in "1997" quickly should highlight 1997
  - Current functionality: Typing in "1997" quickly highlights 1999 after "19" has been typed, but "1997" is not selected after finishing up with "97". Trying to type in "1997" again leads to 1998, then 97, 96 and so on being highlighted, until it reaches 1990.

The plugin defaults example [here](https://tanvir0604.github.io/jquery-dropdown-datepicker/) follows a lot of the UX that I was expecting, though tab does not switch fields when the dropdown is open.

When the month/year fields are focused using tab, the highlighted box does not match the colored section of the underline:
![image](https://user-images.githubusercontent.com/20056751/149824848-a1618a20-2e27-4b0a-bc49-2e11b0dbc8bc.png)

Similarly, having no margin to the left of the text looks a little off balance, especially with Day and some of the months:
![image](https://user-images.githubusercontent.com/20056751/149825047-46dd4e8f-104c-487e-b871-6dd6a27cf05a.png)

My initial thought was that the highlighted option in the drop down was a bit too low contrast when compared to the white background, but it looks better now that I've been playing around with it for a while. Could be an issue on lower contrast monitors though, but should be fine.
![image](https://user-images.githubusercontent.com/20056751/149825386-0ff8e81b-16af-471c-a62d-607df892ac97.png)

marta- committed:
CARDS-1561: On the patient identification screen, replace the date field with dropdowns for day,month,year
Addressing review comments: aligned the highlighting background with the highlighting underline

marta- commented:
The keyboard interaction issues have been reported separately as https://phenotips.atlassian.net/browse/CARDS-1578.

The highlighting alignment issues have been addressed in 41925c9.

sashaandjic reviewed (approved):
re-tested

sdumitriu reviewed (approved):
I really wish keyboard interaction worked.
